### DISCUSSION TITLE
Is it possible to implement an Internet connection check?

### DISCUSSION THREAD

**Discussion by skyvladik** [arendst/Tasmota#17195] [Q&A] [answered]
Hello! Is it possible to implement an Internet connection check? I want to reboot the router.

**Comment by barbudor** [comment, marked as answer]
Indeed, you have choice of using WebQuery or Ping to find out if a server outside is out of reach, if so POWER OFF the relay (having Pulsetime and PowerOnState 5 will help to have the power switched back automatically after X seconds)

Exemple with ping
```
Pulsetime 50
PowerOnState 5
Rule1 on Time#Minute|5 do backlog var1 0;ping4 8.8.8.8;ping4 1.1.1.1;ping4 208.67.222.222; RuleTimer1 60 endon
          on Ping#8.8.8.8#Reachable=true do var1 1 endon 
          on Ping#1.1.1.1#Reachable=true do var1 1 endon 
          on Ping#208.67.222.222#Reachable=true do var1 1 endon 
          on Rules#Timer=1 do Power %var1% endon
Rule1 1
```
The rule has been splitted on multiple lines for readability but of course it should be entered as a single one

Pinging 3 different servers (Google DNS, Cloudflare DNS and OpenDNS). If at least one respond, the power off will not be triggered

If you router takes more than 3 minutes to get back on line, you'd should increase the periodicity of testing

Edited to fix the typo reported by @sfromis 👍

**Reply by sfromis** [reply]
Typo.... Last command should be `Rule1 1`

The rules cookbook also has examples of router watchdog
https://tasmota.github.io/docs/Rules/#watchdog-for-wi-fi-router-or-modem

**Reply by andypnz** [reply]
Hi. New to Tasmota. Built and installed firmware including ping (ESP8266). I added the above in configuration, other, template. I have a couple of questions please:
1. How to I select which GPIO will work my relay (to power cycle the router)
2. Can I add something to the console view so I can see the info on the pings?

**Reply by sfromis** [reply]
Tasmota already publishes Ping results, visible in the console. The rules above are actually based on what you see there.
You select gpio by configuring it as a relay, when configuring the template, and then the Power command has an index, like Power2 for the 2nd relay. Default is index 1.

**Reply by andypnz** [reply]
Thank you. Using a wemos D1 mini + relay on D1 (GPIO5). I added ping to the firmware (and I can ping an external IP address from the console). And I have also added #define USE_RULES.

I copied the above code and pasted it into the console. I now see the 3 pings happening every 5 minutes.

Under 'config template' I have set GPIO5 to Relay, 1. Toggle On/Off turns the relay on and off. Is that the correct config for the relay? I guess because it says 'Power 1' then 'Relay 1' is the correct relay, right?

Thanks again
```
06:00:04.785 RSL: RESULT = {"Ping":{"8.8.8.8":{"Reachable":true,"IP":"8.8.8.8","Success":4,"Timeout":0,"MinTime":62,"MaxTime":99,"AvgTime":73}}}
06:00:04.805 RUL: PING#8.8.8.8#REACHABLE=TRUE performs "var1 1"
06:00:04.810 RSL: RESULT = {"Var1":"1"}
06:00:05.034 RSL: RESULT = {"Ping":{"1.1.1.1":{"Reachable":true,"IP":"1.1.1.1","Success":4,"Timeout":0,"MinTime":48,"MaxTime":62,"AvgTime":54}}}
06:00:05.057 RUL: PING#1.1.1.1#REACHABLE=TRUE performs "var1 1"
06:00:05.062 RSL: RESULT = {"Var1":"1"}
06:00:05.285 RSL: RESULT = {"Ping":{"208.67.222.222":{"Reachable":true,"IP":"208.67.222.222","Success":4,"Timeout":0,"MinTime":64,"MaxTime":91,"AvgTime":75}}}
06:00:05.311 RUL: PING#208.67.222.222#REACHABLE=TRUE performs "var1 1"
06:00:05.315 RSL: RESULT = {"Var1":"1"}
06:01:01.380 RUL: RULES#TIMER=1 performs "Power 1"
06:01:01.385 RSL: RESULT = {"POWER":"ON"}
06:01:01.388 RSL: POWER = ON
```

**Reply by sfromis** [reply]
Yes, as I said, default is index 1 when you have no index. You'd get the same if explicitly having the index, like `Power1 1`

**Reply by andypnz** [reply]
Thanks!

**Reply by mikeymileos** [reply]
My plug doesn't have Ping flashed onto it, but WebQuery works though I can't find much documentation on how to use WebQuery in the same way as Ping to return a result in line to use a solution like the one in the answer.

I want to basically check 3 times if it can't connect and then cycle the power

I don't know if ON events are sequential or what, so if they're not run sequentially, this won't work, but if they are..

Will this work?

Rule1
	ON system#boot DO backlog PowerOnState 5; Pulsetime 115 ENDON
	ON Time#Minute|5 DO WebQuery http://google.gr GET ENDON
	ON WebQuery#Data$!Done DO WebQuery http://google.gr GET ENDON
	ON WebQuery#Data$!Done DO WebQuery http://google.gr GET ENDON
	ON WebQuery#Data$!Done DO Power1 0 ENDON
	ON WebQuery#Data=Done DO Power1 1 ENDON


p.s. I put the PowerOnState in the Rule so if the power goes out it resets that PowerOnState on reboot as well
And I put the last line just incase it's somehow off

I started playing with my first Tasmota programming a few hours ago, and this fun :D (I used to be a programmer though for many years)

Thanks!

**Reply by mikeymileos** [reply]
UPDATE: This didn't seem to work. Though I'm running it in the WebUI Console and maybe it's not verbose, but it only showed one WebQuery failed after I pulled out my telephone cable (I have it on a 1 min timer though for testing)

Is there a way to do multiple WebQueries and update a variable based on them, like the example in the answer above? 

Thanks


UPDATE:

I actually changed it to just poll every minute checking the variable to see if it's past a certain count of fails, and if it fails, add 1 to the variable. If it doesn't fail it sets it to 0.

Rule1
	ON system#boot DO backlog Var1=0; Var2=5; PowerOnState 5; Pulsetime 115 ENDON
	ON Time#Minute|%Var2% DO IF (Var1<3) WebQuery http://google.gr GET ELSE backlog Var2=5; Var1=0; Power1 0 ENDIF ENDON
	ON WebQuery#Data$!Done DO backlog Var1=Var1+1; Var2=1 ENDON
	ON WebQuery#Data=Done DO backlog Var1=0; Var2=1; ENDON
	

Edit: I forgot to take into account modem reboot time. So I updated it to set the Timer every 5 min on Reboot of the plug itself and then if it fails or succeeds in the WebQuery it makes it 1 min. If it turns off the modem, it sets the time to 5 first

**Comment by skyvladik** [comment]
Thanks! Ready-made firmware will not go? Do I need to download the source code and build the firmware?

**Reply by sfromis** [reply]
The WebQuery command used in one of the cookbook examples from my link is available with standard precompiled binaries, unlike using the Ping command.

**Reply by skyvladik** [reply]
For me, WebQuery is very difficult, I can't figure it out.
